hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/ConcurrentArrayRingbuffer.java

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.metrics.managementcenter;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static java.lang.System.arraycopy;
+
+public class ConcurrentArrayRingbuffer<E> {
+
+    private static final Object[] EMPTY_ARRAY = {};
+
+    private final E[] ringItems;
+    /**
+     * Head is the sequence where oldest item is found
+     */
+    private long head;
+    /**
+     * Tail is the sequence where next item will be added
+     */
+    private long tail;
+    private final int capacity;
+
+    @SuppressWarnings("unchecked")
+    public ConcurrentArrayRingbuffer(int capacity) {
+        this.capacity = capacity;
+        this.ringItems = (E[]) new Object[capacity];
+    }
+
+    /**
+     * Appends an item to the ring buffer. If size == capacity, also drops the
+     * oldest item.
+     */
+    public synchronized void add(E item) {
+        if (tail - capacity == head) {
+            head++;
+        }
+        ringItems[toIndex(tail++)] = item;
+    }
+
+    public synchronized void clear() {
+        Arrays.fill(ringItems, null);
+        head = tail;
+    }
+
+    public synchronized E get(long sequence) {
+        checkSequence(sequence);
+        return ringItems[toIndex(sequence)];
+    }
+
+    /**
+     * Copies all the items from the given {@code sequence} up to the
+     * tail. If the item at the sequence number is already dropped, start from
+     * the oldest item.
+     *
+     * @throws IllegalArgumentException If the sequence is in the future.
+     */
+    public synchronized RingbufferSlice<E> copyFrom(long sequence) {
+        sequence = Math.max(sequence, head);
+        if (sequence == tail) {
+            return new RingbufferSlice(EMPTY_ARRAY, tail);
+        }
+        checkSequence(sequence);
+        E[] result = (E[]) new Object[(int) (tail - sequence)];
+        int startPoint = toIndex(sequence);
+        int endPoint = toIndex(tail);
+        if (startPoint >= endPoint) {
+            arraycopy(ringItems, startPoint, result, 0, capacity - startPoint);
+            arraycopy(ringItems, 0, result, capacity - startPoint, endPoint);
+        } else {
+            arraycopy(ringItems, startPoint, result, 0, endPoint - startPoint);
+        }
+        return new RingbufferSlice<>(result, tail);
+    }
+
+    public synchronized int getCapacity() {
```

Comment:
Does this method have to be synchronized, considering that `capacity` is `final`?

Reply:
Feels a copy-paste leftover. Removed synchronized.